src/Lean/ReducibilityAttrs.lean

```diff
@@ -65,15 +70,75 @@ def setReducibilityStatusCore (env : Environment) (declName : Name) (status : Re
 private def setReducibilityStatusImp (env : Environment) (declName : Name) (status : ReducibilityStatus) : Environment :=
   setReducibilityStatusCore env declName status .global .anonymous
 
+/-
+TODO: it would be great if we could distinguish between the following two situations
```

Comment:
Could `ReducibilityHints.abbrev` be used to distinguish between these cases?

Reply:
Unfortunately, it does not help. The key issue here is that the current attribute management system does not tell us whether an attribute is being set at a declaration time or not.

Reply:
True, having the attribute know whether it is applied at declaration time would be the best solution.  However, I was proposing a potential half-baked solution assuming that was too complicated for now. 

If I understand correctly, `abbrev` is just a ``@[reducible, inline] def`` with `ReducibilityHints.abbrev`. Thus, the primary use case for a declaring `@[reducible] def` over an `abbrev`  is when the definition is meant to be reducible but not inlined. In such a case,  `@[noinline] abbrev` could serve the same purpose. Thus, to avoid the potential for applying `@[reducible]` after a declaration, we could forbid `@[reducible]` from being applied to anything except an `abbrev`. This would avoid the need to make the declaration time distinction.

Reply:
@tydeu This is a good suggestion. Then, we would not even have the attribute `[reducible]`. It seems there are very few instances of `@[reducible] def` in Mathlib. So, the impact would be minimal. @semorrison What do you think? We can use a different PR for this change.

Reply:
Fwiw, adding a flag to `AttributeImpl.add` that tells us whether this is a post-hoc application seems like a simple (apart from downstream breakage) and generally very useful change to me